packages/swingset-liveslots/src/cache.js

```diff
@@ -0,0 +1,100 @@
+import { Fail } from '@agoric/assert';
+
+/**
+ * @template V
+ * @callback CacheGet
+ * @param {string} key
+ * @returns {V}
+ */
+
+/**
+ * @template V
+ * @callback CacheSet
+ * @param {string} key
+ * @param {V} value
+ * @returns {void}
+ */
+/**
+ * @callback CacheDelete
+ * @param {string} key
+ * @returns {void}
+ *
+ * @callback CacheFlush
+ * @returns {void}
+ *
+ * @callback CacheInsistClear
+ * @returns {void}
+ */
+/**
+ * @template V
+ * @typedef {object} Cache
+ * @property {CacheGet<V>} get
+ * @property {CacheSet<V>} set
+ * @property {CacheDelete} delete
+ * @property {CacheFlush} flush
+ * @property {CacheInsistClear} insistClear
+ */
+
+/**
+ * Cache of virtual object/collection state
+ *
+ * This cache is empty between deliveries. Within a delivery, the
+ * first access to some data will cause vatstore reads to populate the
+ * cache, then the data is retained until end-of-delivery. Writes to
+ * data will update the cache entry and mark it as dirty. At
+ * end-of-delivery, we flush the cache, writing out any dirty entries,
+ * and deleting all entries.
+ *
+ * This needs RAM for everything read during a delivery (rather than
+ * having a fixed maximum size), but yields a simple (easy to debug)
+ * deterministic relationship between data access and reads/writes to
+ * the backing store.
+ *
+ * @template V
+ * @param {(key: string) => V} readBacking
+ * @param {(key: string, value: V) => void} writeBacking
+ * @param {(key: string) => void} deleteBacking
+ * @returns {Cache<V>}
+ *
+ * This cache is part of the virtual object manager and is not intended to be
+ * used independently; it is exported only for the benefit of test code.
+ */
+export function makeCache(readBacking, writeBacking, deleteBacking) {
```

Comment:
This all seems quite reasonable, but I'm inclined to think that it should be given a name that's somewhat less generic.  It's a specific kind of cache with specific and somewhat unusual semantics, so perhaps it might be better if the name reflected that.

Reply:
`makeFlushableCache`?

Reply:
The analog in [cosmic-swingset/src/helpers/bufferedStorage.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/cosmic-swingset/src/helpers/bufferedStorage.js) is `makeReadCachingStorage` wrapping a kvStore with a read cache and wrapping _that_ with `makeBufferedStorage` (which holds writes until a `commit()` or `abort()`). With the understanding that _this_ cache is based around functions rather than object methods, agreement on vocabulary would still be nice.

Reply:
I can't think of anything better than `makeCache` right now.. if this were exported for general use, where the name must sit amongst other types of caches, it would certainly need a more distinctive name. But while it's specific to liveslots, it can just be "the `cache` that liveslots uses". So I'm gonna leave the name change for another day.